src/main/java/nextstep/subway/line/domain/Sections.java

```diff
@@ -2,29 +2,103 @@
+    private List<Section> sections = new ArrayList<>();
 
-	public void add(Section section) {
-		sections.add(section);
-	}
+    public void add(Section section) {
+        if (sections.isEmpty()) {
+            sections.add(section);
+            return;
+        }
 
-	public List<Station> stations() {
-		List<Station> stations = new ArrayList<>();
-		for (Section section : sections) {
-			stations.addAll(Arrays.asList(section.getUpStation(), section.getDownStation()));
-		}
-		return stations;
-	}
+        validate(section);
+
+        if (isFirstSection(section)) {
+            sections.add(FIRST_INDEX, section);
+            return;
+        }
+        if (isLastSection(section)) {
+            sections.add(section);
+            return;
+        }
+
+        Direction direction = isUpDirection(section) ? Direction.UP : Direction.DOWN;
+        changeSection(section, direction);
+    }
+
+    private boolean isLastSection(Section section) {
+        int lastIndex = sections.size() - COMPLEMENT;
+        return sections.get(lastIndex).getDownStation() == section.getUpStation();
+    }
+
+    private boolean isFirstSection(Section section) {
+        return sections.get(FIRST_INDEX).getUpStation() == section.getDownStation();
+    }
+
+    private boolean isUpDirection(Section section) {
+        return sections.stream().anyMatch(s -> s.getUpStation() == section.getUpStation());
+    }
+
+    private void changeSection(Section section, Direction direction) {
+        sections.stream()
+                .filter(s -> s.getStation(direction) == section.getStation(direction))
+                .findFirst()
+                .ifPresent(s -> {
+                    int distance = s.getDistance() - section.getDistance();
+                    Section moveSection = createByDirection(direction, section, s, distance);
+                    moveSection.toLine(section.getLine());
+
+                    int index = sections.indexOf(s);
+                    sections.remove(s);
+
+                    List<Section> asList = direction == Direction.UP ?
+                        Arrays.asList(section, moveSection) : Arrays.asList(moveSection, section);
+                    this.sections.addAll(index,asList);
+                });
+    }
+
+    private Section createByDirection(Direction direction, Section newSection, Section oldSection, int distance) {
+        if (direction == Direction.UP) {
+            return new Section(newSection.getStation(Direction.DOWN), oldSection.getStation(Direction.DOWN), distance);
+        }
+        return new Section(oldSection.getStation(Direction.UP), newSection.getStation(Direction.UP), distance);
+    }
+
+    public List<Station> stations() {
+        Set<Station> stations = new LinkedHashSet<>();
+        for (Section section : sections) {
+            stations.addAll(Arrays.asList(section.getUpStation(), section.getDownStation()));
+        }
+        return new ArrayList<>(stations);
```

Comment:
```suggestion
        return sections.stream()
                .flatMap(section -> Stream.of(section.getUpStation(), section.getDownStation()))
                .collect(Collectors.toSet());
```
혹은
```suggestion
        return sections.stream()
                .flatMap(section -> Stream.of(section.getUpStation(), section.getDownStation()))
                .distinct();
                .collect(Collectors.toList());
```
이렇게 표현할 수 있지 않을까요?